scRNA-seq/04-dimension_reduction_scRNA.Rmd

```diff
@@ -31,226 +42,294 @@ library(magrittr)
-Use the `cell_qc_df` data frame we just created, to make some plots and decide on cutoffs!
+### Filtering with `emptyDrops()`
+
+The `DropletUtils` package that we used to read in the 10X data has a number of other useful features. 
+One is the `emptyDrops()` function, which uses the overall gene expression patterns in the sample to identify droplets that are likely to not contain an intact cell, but may simply have contained loose ambient RNA released during cell separation. 
+
+This method, described in [Lun *et al.* (2019)](https://doi.org/10.1186/s13059-019-1662-y), uses the droplets with very low UMI counts to estimate the "ambient" expression pattern of RNA content outside of cells.
+It then scores the remaining cells based how much they deviate from that pattern, assigning a small P value when the droplet's expression deviates from the ambient expression pattern.
+Because it uses the low UMI count droplets, this method should not be used when other filtering has already been performed (which is unfortunately the case with the version of `salmon alevin` we used).
+
+This method seems to perform well to exclude false "cells" while retaining cells with distinct expression profiles but low counts that might have failed a simple cutoff. 
+Note that this method also requires that the data has already been quantified, with reads assigned to genes, as compared to a simple total UMI count filter which can be performed much earlier in the pipeline.
 
-```{r qc_plots, live = TRUE}
-# Plot some aspect of the QC data stored in `cell_qc_df`
-ggplot(cell_qc_df, 
-       aes(x = sum, y = detected, color = subsets_mito_percent)) +
-  geom_point() +
-  scale_color_viridis_c() +
-  scale_x_log10() +
-  scale_y_log10() +
-  geom_hline(yintercept = 200, color = "red")
+The `emptyDrops()` function takes the `counts` matrix from our SingleCellExperiment, and returns a data frame with the statistics it calculated.
+This will take a few minutes to run, but we can speed it up by allowing parallel processing.
+
+```{r emptydrops, live = TRUE}
+droplet_stats <- DropletUtils::emptyDrops(
+  counts(hodgkins_sce),
+  BPPARAM = BiocParallel::MulticoreParam(4)) # use multiprocessing
 ```
 
-```{r qc_plot2, live = TRUE}
-# make another QC plot
-ggplot(cell_qc_df, 
-       aes(x = subsets_mito_percent)) +
-  geom_histogram() + 
-  geom_vline(xintercept = 20, color = "red")
+We will use a false discovery rate (FDR) of 0.01 as our cutoff for "real" cells. 
+Since `emptyDrops()` uses low count cells to estimate the "ambient" expression pattern, those cells are not assigned an FDR value, and have a value of NA.
+These NAs can be a problem for filtering with a Boolean vector, as we did above, so instead we will use the `which()` function to get the *positions* of the cells that pass our filter and select the columns we want using that.
+
+```{r filter_empty, live = TRUE}
+cells_to_retain <- which(droplet_stats$FDR <= 0.01)
```

Comment:
Comment for when it gets presented: would make sure to take a look at `cells_to_retain` the global environment